### PR TITLE
Added tests in the style of test-vectors for credential creation

### PR DESCRIPTION
**Pull request by andresuribe87**
# Overview
This is in preparation before promoting them to the `sdk-development` repo.

# Description
This is part of our push to have test vectors. As a result of adding them, I had to add validation to the `fromJson` function. This is the first step in the direction to have test vectors for verifiable credentials 1.1. Heavily influenced from https://github.com/w3c/vc-test-suite/blob/gh-pages/test/vc-data-model-1.0/input/example-017-missing-verifiableCredential.jsonld

# How Has This Been Tested?
This is simply adding more testing area.

### CODE REVIEW

File: credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt [hunk 1]
```diff
@@ -306,7 +309,108 @@ public class VerifiableCredential internal constructor(public val vcDataModel: V
...
      */
     public fun fromJson(vcJson: String): VerifiableCredential {
-      return VerifiableCredential(VcDataModel.fromJson(vcJson))
+      val typeRef = object : TypeReference<HashMap<String, Any>>() {}
+      val vcMap = objectMapper.readValue(vcJson, typeRef)
+      validateDataModel(vcMap)
+      return VerifiableCredential(VcDataModel.fromMap(vcMap))
+    }
+
+    private fun validateDataModel(model: Map<String, Any>) {
```
Comment (amika-sq) [inline credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt:318]: Should we be using the `@Throws` annotation here? Seems like there's two possible outcomes for this function:
1. The function completes
2. The function throws

That didn't really occur to me on my first read-through that it'd throw. Would be great to add for clarity for readers!

File: credentials/src/test/kotlin/web5/sdk/credentials/StatusListCredentialTest.kt [hunk 2]
```diff
@@ -27,17 +27,8 @@ class StatusListCredentialTest {
...
     )
 
-    assertTrue(
-      specExampleRevocableVc.vcDataModel.contexts.containsAll(
-        listOf(
-          URI.create("https://www.w3.org/ns/credentials/v2"),
-          URI.create("https://www.w3.org/ns/credentials/examples/v2")
-        )
-      )
-    )
```
Comment (amika-sq) [inline credentials/src/test/kotlin/web5/sdk/credentials/StatusListCredentialTest.kt:37]: Should this be removed? Seems unrelated to the test vectors on first glance.

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt [hunk 3]
```diff
@@ -94,40 +101,75 @@ class VerifiableCredentialTest {
...
+      """{
+          "kty": "EC",
+          "d": "OeBCt3M8roz9F9Ny192TtZXf_-qPzAkwSM6ep7QdSO4",
+          "use": "sig",
+          "crv": "secp256k1",
+          "kid": "NB_qlUIr06-AikVVMFreq0lc-omQtzc6lwhhcvgO6r4",
+          "x": "DdtN8W6x_34pB_nkxR0e1tmDkNnsJeusBAEPzKWgf_Y",
+          "y": "u3W135inodLqtcEb9jNGS3JsM_uFKmkJSb8Trc9luWI",
+          "alg": "ES256K"
+        }"""
```
Comment (amika-sq) [inline credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt:115]: I'm not seeing this being used anywhere. Still needed?

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt [hunk 4]
```diff
@@ -94,40 +101,75 @@ class VerifiableCredentialTest {
...
+      assertFails(vector.description) {
+        VerifiableCredential.fromJson(mapper.writeValueAsString(testInput.credential))
+      }
+    }
+  }
+
+  @Test
+  fun foo() {
+    assertFalse(URI("example").isAbsolute)
   }
```
Comment (amika-sq) [inline credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt:173]: Still needed?

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt [hunk 5]
```diff
@@ -94,40 +101,75 @@ class VerifiableCredentialTest {
...
+    }
 
-    assertEquals(holderDid.uri, vcDataModel.credentialSubject.id.toString())
-    assertContains(vcDataModel.types, "StreetCred")
+    testVectors.vectors.filter { it.errors ?: false }.forEach { vector ->
+      val testInput = vector.input
+      assertFails(vector.description) {
+        VerifiableCredential.fromJson(mapper.writeValueAsString(testInput.credential))
+      }
+    }
```
Comment (amika-sq) [inline credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt:167]: Just a thought on structure here:

We could have two test vectors: `create-success.json` and `create-failure.json`. That way we could have separate functions that test success/failure scenario scenarios instead of filtering on the inputs. 

But I'm ok with this too! I'm sure our test vectors will evolve over time, so definitely not blocking by any means.

File: dids/src/main/kotlin/web5/sdk/dids/extensions/DidExtensions.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import web5.sdk.dids.methods.key.DidKey
+import web5.sdk.dids.methods.web.DidWeb
+
+internal val supportedMethods = mapOf(
+  DidKey.methodName to DidKey.Companion,
+  DidJwk.methodName to DidJwk.Companion,
+  DidIon.methodName to DidIon.Default,
+  DidDht.methodName to DidDht.Default,
+  DidWeb.methodName to DidWeb.Default
+)
```
Comment (amika-sq) [inline dids/src/main/kotlin/web5/sdk/dids/extensions/DidExtensions.kt:18]: This looks like the same values as `methodResolves` in the `DidResolvers.kt` class. Wondering if it might make sense to combine them?

**Review by amika-sq — APPROVED** [review]
Some comments, but nothing blocking from me. Looking forward to having more test vectors!

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt [hunk 7]
```diff
@@ -94,40 +101,75 @@ class VerifiableCredentialTest {
...
+  data class TestInput(
+    val signerDidUri: String?,
+    val signerPrivateJwk: Map<String, Any>?,
+    val credential: Map<String, Any>?,
+  )
 
-    val vcDataModelJson = jwtClaims.getJSONObjectClaim("vc")
-    val vcDataModel = VcDataModel.fromMap(vcDataModelJson)
+  @Test
+  fun `creates test vectors`() {
```
Comment (KendallWeihe) [inline credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt:143]: Slightly misleading name, in that, it could be interpreted that this function creates the test vectors JSON

Maybe...

```suggestion
  fun `test vectors: creates`() {
```

**Comment by KendallWeihe** [comment]
Is our `validateDataModel` based on a spec? Or is it a TBD prescription? 

If it's an encoding of a spec, could we link to the spec?

File: credentials/src/test/kotlin/web5/sdk/credentials/StatusListCredentialTest.kt (hunk 2 above)
Comment (andresuribe87) [reply credentials/src/test/kotlin/web5/sdk/credentials/StatusListCredentialTest.kt:37]: I removed it because we were creating a VC that was following the 2.0 model, but we're implementing the 1.1 version of VCs. 

It goes hand in hand with the changes to the `revocable_vc.json` file.

File: credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt (hunk 1 above)
Comment (andresuribe87) [reply credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt:326]: Excellent point, added comments. Let me know if it's more clear now.

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt (hunk 3 above)
Comment (andresuribe87) [reply credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt]: It wasn't, thanks for catching! Removed

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt (hunk 4 above)
Comment (andresuribe87) [reply credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt]: Removed.

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt (hunk 7 above)
Comment (andresuribe87) [reply credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt]: Unforturnately, cannot include `:` chars in the test names. Instead, I went with the approach of defining a new class as shown below. Open to suggestions/feedback!

```kt
class TestVectorsCredentialsTest {
  @Test
  fun creates() {
// ...
  }
}
```

File: credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt (hunk 5 above)
Comment (andresuribe87) [reply credentials/src/test/kotlin/web5/sdk/credentials/VerifiableCredentialTest.kt]: I like that! Used `_` instead of `-` because it's what was allowed :)

File: dids/src/main/kotlin/web5/sdk/dids/extensions/DidExtensions.kt (hunk 6 above)
Comment (andresuribe87) [reply dids/src/main/kotlin/web5/sdk/dids/extensions/DidExtensions.kt:18]: Combined!

**Comment by andresuribe87** [comment]
> Is our `validateDataModel` based on a spec? Or is it a TBD prescription?
> 
> If it's an encoding of a spec, could we link to the spec?

@KendallWeihe I added a comment clarifying and pointed to the spec. Let me know if that's clear now.

File: credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt (hunk 1 above)
Comment (nitro-neal) [reply credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt:326]: This could live in a new validators file

Comment (andresuribe87) [reply credentials/src/main/kotlin/web5/sdk/credentials/VerifiableCredential.kt:326]: I would rather keep it private unless it's necessary to make this function public. Do you have a use-case in mind that requires it to be public?
